src/FtpExtractor.php

```diff
@@ -0,0 +1,121 @@
+     */
+    private $filesToDownload;
+
+    public function __construct(bool $onlyNewFiles, bool $isWildcard, FtpFilesystem $ftpFs)
+    {
+        $this->ftpFilesystem = $ftpFs;
+        $this->onlyNewFiles = $onlyNewFiles;
+        $this->isWildcard = $isWildcard;
+        $this->filesToDownload = [];
+    }
+
+    public function copyFiles(string $sourcePath, string $destionationPath, FileStateRegistry $registry): int
+    {
+        if ($this->isWildcard) {
+            $this->prepareToDownloadFolder($sourcePath, $destionationPath);
+        } else {
+            $this->prepareToDownloadSingleFile($sourcePath, $destionationPath);
+        }
+        return $this->download($registry);
+    }
+
+    private function prepareToDownloadFolder(string $sourcePath, string $destinationPath): void
+    {
+        $basePath = Glob::getBasePath(GlobValidator::convertToAbsolute($sourcePath));
+        try {
+            $items = $this->ftpFilesystem->listContents($basePath, self::RECURSIVE_COPY);
+        } catch (\LogicException $e) {
+            throw new UserException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        foreach ($items as $item) {
+            if ($this->isWildcard && !GlobValidator::validatePathAgainstGlob($item['path'], $sourcePath)) {
+                continue;
+            }
+
+            if ($item['type'] === self::FTP_FILETYPE_FILE) {
+                $this->prepareToDownloadSingleFile($item['path'], $destinationPath);
+            }
+        }
+    }
+
+    private function prepareToDownloadSingleFile(string $sourcePath, string $destinationPath): void
+    {
+        $destination = $destinationPath . '/' . strtr($sourcePath, ['/' => '-']);
+        try {
+            $timestamp = (int) $this->ftpFilesystem->getTimestamp($sourcePath);
+        } catch (FileNotFoundException $e) {
+            throw new UserException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $this->filesToDownload[] = [
+            self::FILE_DESTINATION_KEY => $destination,
+            self::FILE_SOURCE_KEY => $sourcePath,
+            self::FILE_TIMESTAMP_KEY => $timestamp,
+        ];
+    }
+
+    private function download(FileStateRegistry $registry): int
+    {
+        $cbTimestampSort = function (array $a, array $b) {
+            return intval($a[self::FILE_TIMESTAMP_KEY]) > intval($b[self::FILE_TIMESTAMP_KEY]);
+        };
+        uasort($this->filesToDownload, $cbTimestampSort);
+
+        $fs = new Filesystem();
+        $downloadedFiles = 0;
+        foreach ($this->filesToDownload as $file) {
+            if ($this->onlyNewFiles
+                && !$registry->shouldBeFileUpdated(
+                    $file[self::FILE_SOURCE_KEY],
+                    $file[self::FILE_TIMESTAMP_KEY]
+                )
+            ) {
+                continue;
+            }
+            $fs->dumpFile(
+                $file[self::FILE_DESTINATION_KEY],
+                $this->ftpFilesystem->read($file[self::FILE_SOURCE_KEY])
```

Comment:
Unhandled \League\Flysystem\FileNotFoundException - kdyz to zmizi mezi vylistovanim a stazenim

Reply:
Jasná věc. Nějak mi storma neznázorňovala neošetřený vyjímky :-1:

Reply:
![image](https://user-images.githubusercontent.com/642928/48342307-96c43700-e66f-11e8-8150-b182e89f1868.png)

Reply:
Díky Tome:) Je vidět, že PhpBouřku máš opravdu zmáklou.

Reply:
https://pehapkari.cz/kurz/ovladni-php-storm-od-zakladu-po-profi-tipy/ ;)